### PR TITLE
fix(#4157): useSpinButton: rapidly changing values causes live region to announce values that are no longer relevant

### LINKED ISSUE #4157: useSpinButton: rapidly changing values causes live region to announce values that are no longer relevant.
# 🐛 Bug Report 

## 🤔 Expected Behavior

Using a screen reader, if you press and hold ArrowUp or ArrowDown key to quickly increment or decrement the value, releasing the key to stop changing the value should only announce the current value.  Instead, the live region announcer assertively announces a series of values that have already been passed. For example, using https://react-spectrum.adobe.com/react-spectrum/NumberField.html#example, if one increments up from 1,024 to a value like 1,075, by pressing and holding ArrowUp, the live announcer announces a series of ~23 values from 1,026 through 1,048 and never actually announces the actual new current value.

## 😯 Current Behavior

Using a screen reader, if you press and hold ArrowUp or ArrowDown key to quickly increment or decrement the value, releasing the key to stop changing the value should only announce the new current value. 

## 💁 Possible Solution

Replace: 

https://github.com/adobe/react-spectrum/blob/9eb87968cf943a94676221729c4b7b8c7e80f26e/packages/%40react-aria/spinbutton/src/useSpinButton.ts#L133-L137

with: 

```jsx
  const animationFrameId = useRef<number>(null);
  useEffect(() => {
    if (isFocused.current) {
      if (animationFrameId.current) {
        cancelAnimationFrame(animationFrameId.current);
        animationFrameId.current = null;
        clearAnnouncer('assertive');
      }
      animationFrameId.current = requestAnimationFrame(() => announce(textValue, 'assertive'));
    }
  }, [textValue]);
```

## 🔦 Context

NumberField Accessibility Audit FY23

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| react-spectrum   |  [@adobe/react-spectrum@3.25.0](https://github.com/adobe/react-spectrum/releases/tag/%40adobe%2Freact-spectrum%403.25.0)
| Browser          |  Chrome Version 110.0.5481.177 (Official Build) (arm64)
| Operating System | macOS Ventura 13.2.1 (22D68)

## 🧢 Your Company/Team

Adobe/Accessibility

## 🕷 Tracking Issue (optional)

A11Y-5852

### PR DESCRIPTION
**Pull request by majornista**
Closes #4157 

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue 4157](https://github.com/adobe/react-spectrum/issue/4157).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [x] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:
1. open https://reactspectrum.blob.core.windows.net/reactspectrum/c655ca8987690ae574c5f7dbe3543b59b9e42c93/docs/react-spectrum/NumberField.html#example
2. With a screen reader running, navigate to set focus on the NumberField example with label as "Width" and value 1024.'
3. Press Enter to begin editing the field in forms mode, then press and hold ArrowUp or ArrowDown to increment or decrement the value rapidly by ~20 steps.
4. Screen reader should announce the final value without announcing a bunch of interim values.

## 🧢 Your Project:

Adobe/Accessibility

### CODE REVIEW

File: packages/@react-aria/spinbutton/src/useSpinButton.ts [hunk 1]
```diff
@@ -130,9 +130,15 @@ export function useSpinButton(
   // In addition, replace the empty string with the word "Empty" so that iOS VoiceOver does not read "50%" for an empty field.
   textValue = textValue === '' ? stringFormatter.format('Empty') : (textValue || `${value}`).replace('-', '\u2212');
 
+  const animationFrameId = useRef<number>(null);
   useEffect(() => {
     if (isFocused.current) {
-      announce(textValue, 'assertive');
+      if (animationFrameId.current) {
```
Comment (snowystinger) [inline packages/@react-aria/spinbutton/src/useSpinButton.ts:136]: can we just clear it before the new announce? or does that make a constant stream of weird noises? we're essentially debouncing the announcement?

if we must have the delay, then the raf also needs to be canceled and cleaned up on unmount

Comment (majornista) [reply packages/@react-aria/spinbutton/src/useSpinButton.ts]: It seems to work with just `clearAnnouncer('assertive');`. Thanks!

**Review by snowystinger — APPROVED** [review]
Tried it out with VO, nothing is announced while spinning until the spinning stops, then it only announces the last value. This is much better than previously where it would wait until the spinning stops, then announce every number we'd gone through.

**Comment by majornista** [comment]
FWIW, I experimented with using a `setTimeout` to see if I could prevent announcement of all the interim values with JAWS on Windows, but it did not do enough to alleviate the problem. JAWS seems to throttle execution of the key down event when a user holds down a key, so that each change will be announced. A user can stop this verbose announcement by pressing the Ctrl key to stop JAWS from speaking, and then use the JAWS key (Insert or CapsLock) + Tab to hear the current value of the input.

**Review by LFDanLu — COMMENTED** [review]
Tested in NVDA, VO, and Talkback. Verified that the experience is much better in VO. In NVDA, the prod numberfield docs didn't seem to have the issue described (it announced the final value without the intermediate values) and but still works fine so not sure if I'm doing something wrong there when trying to reproduce. For TalkBack, double tapping and holding one of the increment/decrement buttons still causes an announcement for every intermediate value.

Since this change is a net improvement and doesn't degrade the experience elsewhere I'm fine with it as is. Lemme know how you wanna proceed @majornista

**Comment by majornista** [comment]
@LFDanLu I think the change offers significant enough improvement with VO to be merged. I don't think NVDA had the issue  of all values being announced before.

**Pull request merged**